test failure org.elasticsearch.xpack.dataframe.checkpoint.DataFrameTransformCheckpointServiceNodeTests » testGetCheckpointStats

## Example build failure
https://scans.gradle.com/s/q6iig7skusihw/tests/pzhjuvhaoog6w-fxij42n4odxkk

## Reproduction line 
does not reproduce locally
```
./gradlew :x-pack:plugin:data-frame:test --tests "org.elasticsearch.xpack.dataframe.checkpoint.DataFrameTransformCheckpointServiceNodeTests.testGetCheckpointStats" -Dtests.seed=31B7CC998CFE5D73 -Dtests.security.manager=true -Dtests.locale=fr-MA -Dtests.timezone=Australia/Lindeman -Dcompiler.java=12 -Druntime.java=11
HomeTestsDataFrameTransformCheckpointServiceNodeTests » testGetCheckpointStatsClose test output (esc)
```

## Example relevant log:
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=46, name=elasticsearch[node_s_0][get][T#1], state=RUNNABLE, group=TGRP-DataFrameTransformCheckpointServiceNodeTests]Close stacktrace
at __randomizedtesting.SeedInfo.seed([31B7CC998CFE5D73:1763BD35BF5C2F56]:0)
Caused by: java.lang.AssertionError: got unexpected exception: Failure during source checkpoint info retrievalClose stacktrace
at __randomizedtesting.SeedInfo.seed([31B7CC998CFE5D73]:0)
at org.junit.Assert.fail(Assert.java:88)
at org.elasticsearch.xpack.dataframe.DataFrameSingleNodeTestCase.lambda$assertAsync$2(DataFrameSingleNodeTestCase.java:66)
at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:70)
at org.elasticsearch.action.LatchedActionListener.onFailure(LatchedActionListener.java:50)
at org.elasticsearch.xpack.dataframe.checkpoint.DefaultCheckpointProvider.lambda$getCheckpointingInfo$10(DefaultCheckpointProvider.java:275)
at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:70)
at org.elasticsearch.xpack.dataframe.checkpoint.DefaultCheckpointProvider.lambda$getIndexCheckpoints$4(DefaultCheckpointProvider.java:169)
at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:70)
at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:64)
at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43)
at org.elasticsearch.xpack.dataframe.checkpoint.DataFrameTransformCheckpointServiceNodeTests$MockClientForCheckpointing.doExecute(DataFrameTransformCheckpointServiceNodeTests.java:103)
at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:377)
at org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin(ClientHelper.java:89)
at org.elasticsearch.xpack.core.ClientHelper.executeWithHeadersAsync(ClientHelper.java:152)
at org.elasticsearch.xpack.dataframe.checkpoint.DefaultCheckpointProvider.getIndexCheckpoints(DefaultCheckpointProvider.java:145)
at org.elasticsearch.xpack.dataframe.checkpoint.DefaultCheckpointProvider.lambda$getCheckpointingInfo$11(DefaultCheckpointProvider.java:283)
at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62)
at org.elasticsearch.xpack.dataframe.persistence.DataFrameTransformsConfigManager.parseCheckpointsLenientlyFromSource(DataFrameTransformsConfigManager.java:393)
at org.elasticsearch.xpack.dataframe.persistence.DataFrameTransformsConfigManager.lambda$getTransformCheckpoint$3(DataFrameTransformsConfigManager.java:158)
at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:62)
at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43)
at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:70)
at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:64)
at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:261)
at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:247)
at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1100)
at org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1178)
at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1158)
at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:54)
at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:47)
at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:30)
at org.elasticsearch.action.support.single.shard.TransportSingleShardAction.lambda$asyncShardOperation$0(TransportSingleShardAction.java:110)
at org.elasticsearch.action.ActionRunnable$1.doRun(ActionRunnable.java:45)
at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769)
at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.lang.Thread.run(Thread.java:834)
```

## Frequency

--- a/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java
+++ b/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java
@@ -28,6 +28,7 @@ import org.elasticsearch.xpack.core.dataframe.transforms.DataFrameTransformProgr
 import org.elasticsearch.xpack.dataframe.persistence.DataFrameTransformsConfigManager;
 
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
@@ -136,7 +137,9 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
 
         ClientHelper.executeWithHeadersAsync(transformConfig.getHeaders(), ClientHelper.DATA_FRAME_ORIGIN, client, GetIndexAction.INSTANCE,
                 getIndexRequest, ActionListener.wrap(getIndexResponse -> {
-                    Set<String> userIndices = new HashSet<>(Arrays.asList(getIndexResponse.getIndices()));
+                    Set<String> userIndices = getIndexResponse.getIndices() != null
+                            ? new HashSet<>(Arrays.asList(getIndexResponse.getIndices()))
+                            : Collections.emptySet();
                     // 2nd get stats request
                     ClientHelper.executeAsyncWithOrigin(client,
                         ClientHelper.DATA_FRAME_ORIGIN,

--- a/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java
+++ b/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java
@@ -261,7 +261,7 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
                     listener.onResponse(checkpointingInfoBuilder.build());
                 },
                 e -> {
-                    logger.error((Supplier<?>) () -> new ParameterizedMessage(
+                    logger.debug((Supplier<?>) () -> new ParameterizedMessage(
                             "Failed to retrieve source checkpoint for data frame [{}]", transformConfig.getId()), e);
                     listener.onFailure(new CheckpointException("Failure during source checkpoint info retrieval", e));
                 }
@@ -274,7 +274,7 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
                     getIndexCheckpoints(checkpointsByIndexListener);
                 },
                 e -> {
-                    logger.error((Supplier<?>) () -> new ParameterizedMessage(
+                    logger.debug((Supplier<?>) () -> new ParameterizedMessage(
                             "Failed to retrieve next checkpoint [{}] for data frame [{}]", lastCheckpointNumber + 1,
                             transformConfig.getId()), e);
                     listener.onFailure(new CheckpointException("Failure during next checkpoint info retrieval", e));
@@ -289,7 +289,7 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
                         nextCheckpointListener);
             },
             e -> {
-                logger.error((Supplier<?>) () -> new ParameterizedMessage(
+                logger.debug((Supplier<?>) () -> new ParameterizedMessage(
                         "Failed to retrieve last checkpoint [{}] for data frame [{}]", lastCheckpointNumber,
                         transformConfig.getId()), e);
                 listener.onFailure(new CheckpointException("Failure during last checkpoint info retrieval", e));

--- a/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java
+++ b/x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DefaultCheckpointProvider.java
@@ -261,7 +261,7 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
                     listener.onResponse(checkpointingInfoBuilder.build());
                 },
                 e -> {
-                    logger.debug((Supplier<?>) () -> new ParameterizedMessage(
+                    logger.error((Supplier<?>) () -> new ParameterizedMessage(
                             "Failed to retrieve source checkpoint for data frame [{}]", transformConfig.getId()), e);
                     listener.onFailure(new CheckpointException("Failure during source checkpoint info retrieval", e));
                 }
@@ -274,7 +274,7 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
                     getIndexCheckpoints(checkpointsByIndexListener);
                 },
                 e -> {
-                    logger.debug((Supplier<?>) () -> new ParameterizedMessage(
+                    logger.error((Supplier<?>) () -> new ParameterizedMessage(
                             "Failed to retrieve next checkpoint [{}] for data frame [{}]", lastCheckpointNumber + 1,
                             transformConfig.getId()), e);
                     listener.onFailure(new CheckpointException("Failure during next checkpoint info retrieval", e));
@@ -289,7 +289,7 @@ public class DefaultCheckpointProvider implements CheckpointProvider {
                         nextCheckpointListener);
             },
             e -> {
-                logger.debug((Supplier<?>) () -> new ParameterizedMessage(
+                logger.error((Supplier<?>) () -> new ParameterizedMessage(
                         "Failed to retrieve last checkpoint [{}] for data frame [{}]", lastCheckpointNumber,
                         transformConfig.getId()), e);
                 listener.onFailure(new CheckpointException("Failure during last checkpoint info retrieval", e));

--- a/x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/DataFrameSingleNodeTestCase.java
+++ b/x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/DataFrameSingleNodeTestCase.java
@@ -70,7 +70,7 @@ public abstract class DataFrameSingleNodeTestCase extends ESSingleNodeTestCase {
         }), latch);
 
         function.accept(listener);
-        latch.await(10, TimeUnit.SECONDS);
+        assertTrue("timed out after 20s", latch.await(20, TimeUnit.SECONDS));
     }
 
 }

--- a/x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/checkpoint/DataFrameTransformCheckpointServiceNodeTests.java
+++ b/x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/checkpoint/DataFrameTransformCheckpointServiceNodeTests.java
@@ -72,8 +72,8 @@ public class DataFrameTransformCheckpointServiceNodeTests extends DataFrameSingl
 
     private class MockClientForCheckpointing extends NoOpClient {
 
-        private ShardStats[] shardStats;
-        private String[] indices;
+        private volatile ShardStats[] shardStats;
+        private volatile String[] indices;
 
         MockClientForCheckpointing(String testName) {
             super(testName);
@@ -97,6 +97,7 @@ public class DataFrameTransformCheckpointServiceNodeTests extends DataFrameSingl
 
             if (request instanceof GetIndexRequest) {
                 // for this test we only need the indices
+                assert(indices != null);
                 final GetIndexResponse indexResponse = new GetIndexResponse(indices, null, null, null, null);
 
                 listener.onResponse((Response) indexResponse);
@@ -173,7 +174,6 @@ public class DataFrameTransformCheckpointServiceNodeTests extends DataFrameSingl
                 DataFrameTransformCheckpoint.EMPTY, null, null);
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/45238")
     public void testGetCheckpointStats() throws InterruptedException {
         String transformId = randomAlphaOfLengthBetween(3, 10);
         long timestamp = 1000;

